Refactor: 코드 컨벤션, 네이밍 등 수정

wh5929 opened the pull request:
### 리팩토링

1. 코드 컨벤션 준수
2. UI 객체 프레임 사이즈를 extension 사용하여 변경
3. 클래스, 구조체, 함수, 변수명 변경
4. firebase framework 사용시 필요한 Google...plist 파일 추가

wh5929 committed:
리팩토링

1. 코드 컨벤션 준수
2. UI 객체 프레임 사이즈를 extension 사용하여 변경
3. 클래스, 구조체, 함수, 변수명 변경
4. firebase framework 사용시 필요한 Google...plist 파일 추가

DalioKim commented on GatheRunner/Presentaion/Running/Model/RunGuideExpItem.swift:
```diff
@@ -1,13 +1,13 @@
 //
-//  RunningGuide.swift
+//  RunGuideExpItem.swift
```
[의견] RunGuideExpItem이라는 네이밍이 뭘 나타내는지 알기 어렵습니다.

wh5929 replied:
'러닝 가이드 체험'을 그대로 나열해서 써서
-> RunningGuideExperienceItem 입니다.
클린코드 저자가 이야기하길 위와 같이 긴 변수명은 쓰지 않는것이 좋다고 해서 RunGuideExpItem으로 줄여봤습니다.

그러나 딱보면 딱하고 읽히는 줄임말은 아니었군요.

RunGuideExperienceItem 와 같이 수정해야겠어요.

아래 동일한 리뷰들도 같이 수정하겠습니다.

DalioKim replied:
줄임말은 최대한 지양하셔야 될거에요. 줄임말이 약속된 언어가 아니어서 유추할 수가 없어요.

DalioKim commented on GatheRunner/Presentaion/Running/RunningGuideExperienceTabItem/RunGuideExpTabItem.swift:
```diff
@@ -1,40 +1,40 @@
 //
-//  RunningGuideTabItem.swift
+//  RunGuideExpTabItem.swift
```
[의견] 역시 네이밍이 뭘 나타내는지 알기 어렵습니다.

DalioKim commented on GatheRunner/Presentaion/Running/RunningView.swift:
```diff
@@ -47,9 +47,9 @@ struct RunningView: View {
+// MARK: - MenuBtn
 
-struct commonBtn: View {
+struct MenuBtn: View {
```
[의견] 타입네이밍의 첫글자가 소문자로 잘못되어있던 걸 고친 것과 좀 더 직접적으로 연관된 네이밍으로 바꿔서 더 좋네요.

DalioKim commented on GatheRunner/Presentaion/Running/ViewModel/RunGuideExpItemVM.swift:
```diff
@@ -0,0 +1,54 @@
+
+import SwiftUI
+
+class RunGuideExpItemVM: ObservableObject {
```
[의견] exp가 example의 줄임말인가요? 줄임말은 쓰지 않는게 좋습니다.

DalioKim reviewed (commented):
[의견] PR의 타이틀을 좀더 친절하게 적어주시면 좋을 것 같습니다. 어떤 부분에 대한 리팩토링인지 알 수 없습니다. 
[링크](https://velog.io/@ye-ji/Git-PR-%EC%9E%98-%EC%93%B0%EB%8A%94-%EB%B0%A9%EB%B2%95)

wh5929 committed:
리뷰 내용 반영하여 변수명 수정
Exp -> Experience

wh5929 commented:
> 

확인했습니다.

DalioKim commented on GatheRunner/Presentaion/Running/FloatingView.swift:
```diff
@@ -48,7 +48,7 @@ struct FloatingView: View {
+                        .frame(width: UIScreen.getWidthby(ratio: 0.1154), height: UIScreen.getHeightby(ratio: 0.0592))
                         .font(.system(size: 25)).foregroundColor(Color.black)
                         .background(Color.white)
                         .cornerRadius(200)
```
[의견] 리팩토링을 진행하시는 PR이어서 드리는 의견인떼, floatingView의 버튼들은 추상화와 캡슐화를 한다면 더 코드가 깔금해보일 것 같습니다. 좌우의 설정버튼과 음악어플연결버튼은 이미지를 제외하고는 모두 일치하기때문에 공통의 기능으로 묶는 추상화가 가능할 것 같고, 시작버튼 역시, body가 아닌 부분에서 변수로 캡슐화해주면 더 좋을 것 같습니다.

wh5929 replied:
네 이건 다음 리팩토링때 진행할 예정입니다.

wh5929 replied:
> [의견] 리팩토링을 진행하시는 PR이어서 드리는 의견인떼, floatingView의 버튼들은 추상화와 캡슐화를 한다면 더 코드가 깔금해보일 것 같습니다. 좌우의 설정버튼과 음악어플연결버튼은 이미지를 제외하고는 모두 일치하기때문에 공통의 기능으로 묶는 추상화가 가능할 것 같고, 시작버튼 역시, body가 아닌 부분에서 변수로 캡슐화해주면 더 좋을 것 같습니다.

이 부분 수정할게요

DalioKim commented on GatheRunner/Presentaion/Running/FloatingView.swift:
```diff
@@ -13,14 +13,14 @@ struct FloatingView: View {
                 Button { } label: {
                     Image(systemName: "gearshape.fill")
-                        .frame(width: UIScreen.main.bounds.width * 0.1154, height: UIScreen.main.bounds.height * 0.0592)
+                        .frame(width: UIScreen.getWidthby(ratio: 0.1154), height: UIScreen.getHeightby(ratio: 0.0592))
```
[의견] 구현하시는 부분에서 전체적으로 UI로직에 사용되는 숫자가 리터럴자체로 작성되어 있습니다. 의미도 알기 힘들고, 유지보수도 힘들기때문에 상수로 선언하셔서 구현하시는 편을 추천드립니다. 

[참고링크](https://javabom.tistory.com/28)

wh5929 replied:
확인했습니다. 감사합니다.

DalioKim commented on GatheRunner/Presentaion/Running/ViewModel/RunGuideExperienceItemVM.swift:
```diff
@@ -0,0 +1,54 @@
+
+import SwiftUI
+
+class RunGuideExperienceItemVM: ObservableObject {
```
[의견] 네이밍이 적합하지 않습니다. 뷰모델이란 뷰의 모델인데 특정아이템의 뷰모델로 명명하시는 것은 일반적이지 않은 것 같습니다.

wh5929 replied:
네 RunGuideExperienceVM 으로 수정하는게 좋을 것 같네요.

DalioKim commented on GatheRunner/Presentaion/Running/JustStartView.swift:
```diff
@@ -36,25 +36,18 @@ struct JustStartView: View {
-                        // 드래그할때 딱 한번만 메서드가 실행된다. 따라서 아래 메서드는 드래그 하는 동안 한번만 호출된다.
                         timer.upstream.connect().cancel()
                     })
             .onReceive(timer) { _ in
```
[의견]
1. featuredTabView는 기능과 변수가 많기 때문에 다른 타입으로 분류하는게 적합해보입니다.
2. 네이밍이 적합하지 않다고 생각합니다. 상단의 슬라이드, 가이드 체험의 의미가 포함되면 더 좋을 것 같습니다.
3. 재스쳐와 관련된 메소드들은 따로 캡슐화해주는게 가독성이 더 좋을 것 같습니다. 
4. 컨텡이너뷰격인 RunningView의 selected는 featuredTabView의 프로퍼티로 적합해보입니다. featuredTabView내에서만 사용되고 될것으로 보입니다.

wh5929 replied:
좋은 의견 감사합니다.

2, 3번 수정할게요.

-1. 같은 구조체 내에서 some View 타입 변수로 사용하는 것과, 다른 구조체(다른 타입)로 사용하는것과 어떤 차이가 있을까요?

-4 . 이렇게 설정한 이유는 featuredTabView에서 현재 보여지고 있는 탭의 번호를 저장하고, JustStartView 인스턴스가 해제된 후 다시 실행될 때(바로 시작 뷰 -> 러닝 가이드 뷰 -> 바로 시작 뷰) 같은 탭을 보여주기 위함입니다. 그래서 컨테이너 뷰격인 RunningView에서 해당 프로퍼티를 갖고 있는 방식으로 구현했습니다.

DalioKim replied:
> -1. 같은 구조체 내에서 some View 타입 변수로 사용하는 것과, 다른 구조체(다른 타입)로 사용하는것과 어떤 차이가 있을까요?

이건 제 주관적인 생각이니, 필터링해서 받아들이시면 좋을 것 같아요. 제가 생각하기에 타입(클래스,구조체)이란 특정 역할을 담당하는거라고 생각이 듭니다.

 역할이란 변수와 함수를 통해 어떤 동작들을 하는걸 말하고요.

 featuredTabView를 예로 들어 이야기하면 featuredTabView는 단순히 탭화면만 보여주는게 아니라, 탭화면을 탭하거나 슬라이드하는 부분(또 그 안에서 상세화면 이동, 탭화면 전환, 타이머 초기화등), 타이머 계산을 통한 탭화면 전환 등 여러 역할들을 하고 있다고 생각이 들어서 이 부분은 다른 하나의 타입(역할)로 봐도 무방하다 생각이 들어서 말씀 드렸습니다.

DalioKim replied:
> -4 . 이렇게 설정한 이유는 featuredTabView에서 현재 보여지고 있는 탭의 번호를 저장하고, JustStartView 인스턴스가 해제된 후 다시 실행될 때(바로 시작 뷰 -> 러닝 가이드 뷰 -> 바로 시작 뷰) 같은 탭을 보여주기 위함입니다. 그래서 컨테이너 뷰격인 RunningView에서 해당 프로퍼티를 갖고 있는 방식으로 구현했습니다.

이 부분도 제 주관적인 생각이어서, 알아서 판단하시면 될 것 같습니다.  제 생각에는 

1) 탭 index 유지: featuredTabView에서 state로 인덱스 변수 가지고 있으면 될 것 같습니다.

2) JustStartView에서 featuredTabView의 변수를 사용한다는건, featuredTabView가 수정될때  JustStartView 같이 수정해야할 가능성도 높입니다.

3) 응집도란 한 모듈 내부의 처리 요소들이 서로 관련되어 있는 정도를 말한다. RunningView에서는 selected를 사용할 일이 없고, featuredTabView에서 직접 사용하기 때문에, 응집도 측면에서 봤을때 featuredTabView을 구조체로 선언하고 featuredTabView에서 selected를 state 변수로 선언하는게 좀 더 적합해보입니다.

wh5929 replied:
> 

1. RunGuideTabView(구 featuredTabView) 에서 @State로 인덱스 변수로 갖는 상태로 구현할 경우, 뷰 전환시 인덱스 상태를 보존할 수 없는 문제를 해결하지 못하고 있습니다. 그리고 @State 를 상위뷰에서 @Binding 을 하위뷰에서 사용하는것이 일반적인듯합니다. 따라서 커플링이 발생하더라도 그 정도가 약하므로 RunningGuide.swift 에서 @State를 사용하여 구현하는것에 문제가 없을듯합니다.

이 문제로 꽤 오랜시간을 고민했는데 제 실력으로 다른 방법을 찾기 어려웠네요. 다른 방법으로 커플링을 제거하고 해결하실 수 있으시다면 말씀해주세요.

DalioKim replied:
> 이 문제로 꽤 오랜시간을 고민했는데 제 실력으로 다른 방법을 찾기 어려웠네요. 다른 방법으로 커플링을 제거하고 해결하실 수 있으시다면 말씀해주세요.

개발외적인 이야기인데,  스터디에서는 상관없지만 회사에서 '제 실력으로는' 이런 표현은 다소 자조적이고 부정적으로 보일수도 있을 것 같으니 이런 표현은 안써버릇 하시는게 좋을 것 같아요. 

Swift Ui에서는 매번 이런 상위뷰에 의존하는 문제를 없애기 위해서 환경변수를 사용하는걸로 알고 있는데 환경변수를 이용해서 구현하는건 어떨까요? 
[레퍼런스링크](https://seons-dev.tistory.com/entry/SwiftUI-EnvironmentObject%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EC%97%AC-%EB%B7%B0%EA%B0%84%EC%97%90-%EB%8D%B0%EC%9D%B4%ED%84%B0-%EA%B3%B5%EC%9C%A0)

DalioKim commented:
[질문] RunningGuideView는 왜 따로 타입으로 분류하신건가요?

DalioKim commented on GatheRunner/Presentaion/Running/RunningGuideExperienceTabItem/RunGuideExperienceTabItem.swift:
```diff
@@ -1,40 +1,40 @@
 //
-//  RunningGuideTabItem.swift
+//  RunGuideExperienceTabItem.swift
```
[의견] 다른 서브뷰들과 따로 어떤 차이가 있기에 RunGuideExperienceTabItem만 폴더로 구분을 하신건가요? 서브뷰들끼리 묶거나, 아니면 다른 서브뷰들과 똑같이 기존 폴더경로 밑에 있는게 더 일관성이 있어보입니다.

wh5929 replied:
dalio 가 Running 폴더 하위에 파일들이 너무 많다고 LocationManager, RunGuideTabItem 을 폴더 만들어서 분리하자고 얘기했었습니다. 
신규 TabItem 추가로 해당 폴더에 TabItem 과 관련된 파일이 두개 있는데 해당 폴더 밑에 두고 쓰는것도 괜찮아 보입니다.

wh5929 replied:
> [의견] RunningView의 .ignoresSafeArea(edges: .top) 연산자로 인해 상단의 카테고리버튼이 안됩니다.

네비게이션바가 각 뷰 상단을 덮고 있어 발생한 문제입니다.
MainTabView 에서 .navigationBarHidden(true) 를 적용하면 각 하위 뷰에서 네비게이션바가 각 뷰 상단을 덮는 문제를 해결할 수 있었습니다.

DalioKim replied:
> 네비게이션바가 각 뷰 상단을 덮고 있어 발생한 문제입니다. MainTabView 에서 .navigationBarHidden(true) 를 적용하면 각 하위 뷰에서 네비게이션바가 각 뷰 상단을 덮는 문제를 해결할 수 있었습니다.

해당 부분 따로 브런치 만들어서 수정해서 올리겠습니다.

DalioKim commented on GatheRunner/Presentaion/Running/RunningGuideExperienceTabItem/RunGuideExperienceTabItem.swift:
```diff
@@ -1,40 +1,40 @@
+                    Text(arrsRunGuideExperienceTabItem[index].text)
                         .font(.system(size: 15))
                         .fontWeight(.regular)
                         .foregroundColor(Color.gray)
```
[의견] body부분이 너무 방대해지고 중복적인 코드도 보입니다. 캡슐화와 추상화를 해주심 더 좋을 것 같습니다.

wh5929 replied:
> [의견] body부분이 너무 방대해지고 중복적인 코드도 보입니다. 캡슐화와 추상화를 해주심 더 좋을 것 같습니다.

### 질문
코드의 재사용에 대해서 어떤 기준을 잡아야 할까요?
코드는 길지 않다고 생각하는데, 캡슐화 추상화가 가능한 코드가 있다면 우선은 하는게 좋을까요?
해당 코드가 다른 곳에서 재사용되는 경우가 없고 캡슐화/추상화하려는 대상의 수가 적어서,
캡슐화/추상화하게 되면 결국 같은 기능을 하기 위한 코드 길이의 총합은 비슷할것 같습니다.
코드자체는 한눈에 들어오기 때문에 가독성은 좋다고 생각하구요.
이 부분에 대해 공부를 하고 오겠습니다.

DalioKim replied:
> 코드의 재사용에 대해서 어떤 기준을 잡아야 할까요?

당연히 중복적으로 사용된 코드들을 재사용할 수 있게 추상화하시면 됩니다. 우선 구현하시고나서 보시면 여러 곳에서 중복적으로 사용된 코드를 발견할 수 있는데 그 부분을 추상화하면 저희가 매번 구현할 필요가 없어지죠.

DalioKim replied:
> 코드는 길지 않다고 생각하는데, 캡슐화 추상화가 가능한 코드가 있다면 우선은 하는게 좋을까요?

어떤 부분의 코드가 길지 않다고 생각하시는건가요? 추상화 가능한 부분은 되도록 하는게 좋지않을까 생각합니다만 이 부분은 저도 확실하게 말하긴 어려울 것 같아요.

DalioKim replied:
> 해당 코드가 다른 곳에서 재사용되는 경우가 없고 캡슐화/추상화하려는 대상의 수가 적어서,
> 캡슐화/추상화하게 되면 결국 같은 기능을 하기 위한 코드 길이의 총합은 비슷할것 같습니다.
> 코드자체는 한눈에 들어오기 때문에 가독성은 좋다고 생각하구요.

캡슐화란 객체의 속성과 행위를 하나로 묶고, 실제 구현 내용 일부를 내부에 감추어 은닉하는 것을 말합니다. 

현재 이야기하고 있는 RunningGuideTabItem의 body에서는 ui로직을 모두 보여주고 있죠. 물론 연산자를 통해서 처리한것이기때문에 선언형 프로그래밍이 아니다라는건 아니지만 기본적으로 선언형 프로그래밍이 추구하는 건 로직의 절차를 숨기는 거라고 저는 생각하는데 

그런 측면에서 body에서 각각의 UI요소들을 구성하는 로직을 보여주는건 선언형이 추구하는 것과 다르게, 코드를 보는 입장에서 인지적으로 부하를 준다고 생각듭니다.

지금은 코드를 직접 구현한지 얼마 안되셨으니 눈에 들어오시겠지만, 이걸 구현하지 않은 사람 또는 구현 후 일정시간이 지난뒤보면 
body부분에서 ForEach 반복문을 통해 무언가를 반복적으로 만들어주고, 수평으로 image와 vstack을 배치하고, vstack에서 다시 타이틀과 서브타이틀을 구성하는 부분이라는걸 파악해야하죠. 

하지만 이걸 캡슐화를 통해서 표현하면 hstack에 썸네일과 타이틀스택만 구현하면 되죠. 물론 전체코드의 길이는 기존의 절차형방식이 더 짧습니다. 하지만 우리가 매번 RunningGuideTabItem을 수정할때 UI모든 로직을 볼필요가 없으니까요. body상에는 어떤 뷰들이 있고, 타이틀부분 또는 이미지 부분을 수정해야 한다면 해당 부분만 확인 후 수정할 수 있게되므로 캡슐화를 하는겁니다. 

다시 정리해서 말씀드리면 1.코드는 수정될 가능성이 크다. 2.코드를 다시 수정할때 코드를 보고 이해해야한다. 3.코드를 볼때 한번에 많은 정보가 들어오면 인지적 부하가 심하다. 4.이를 위해서 정보를 숨기고, 범주화(카테고리화)를 통해 이해를 돕고, 수정을 해당 파트만 진행할 수 있게 한다. 

추가적으로 말씀드리면 이러한 캡슐화, 추상화가 없다면 지금 저희가 사용하는 swiftui 프레임워크가 없고 매번 ui를 그릴때마다 지금 간단하게 사용하고 있는 연산자 또는 image text와 같은 객체를 반복적으로 계속 구현해줘야 됩니다. 

마지막으로 코드자체에 가독성이 좋다고 주관적으로 판단하는건  위험하다 생각합니다. 
업무상 이메일, 문서를 작성할때 그 가독성을 판단하는건 작성자가 아니라 그걸 보는 사람이 하는거고,
실제로 가독성 좋은 코드를 작성하는 사람은 정말 극소수입니다. 저희가 현재 하고 있는 이 모든 과정자체가 가독성 좋은 코드를 짜기 위해 연습하는거고요. 저희가 단번에 그런 수준에 오를 수 있다고 생각하는건 정말 위험합니다.

DalioKim replied:
그리고 코드의 가독성의 하나의 기준인 Pyramid of Doom  대해 말씀드릴게요.

vstack {
  hstack {

와 같은 중첩되는 형태의 코드를 말하는데 이는 사람의 언어로 표현하면 누구의 ~ 옷의 단추의~ 구멍의~ 품질의~ 이런식으로 표현되는데 아무래도 그 의미를 이해하기 힘들고, 시각적으로도 피곤함을 줍니다. 

swiftui에서는 클로저를 통해서 ui로직을 구성하다보니 이런 문제가 자주 발생하는데, 최대한 이런 부분을 고려해서 구현해야 가독성 좋은 코드를 작성할 수 있다고 생각이 듭니다.

참고 레퍼런스 입니다.
[Escape the Pyramid of Doom ](https://levelup.gitconnected.com/escape-the-pyramid-of-doom-c58edd326225)
[SOLVED: Suggestions for fixing the pyramid of doom?
](https://www.hackingwithswift.com/forums/swiftui/best-convention-for-fixing-the-pyramid-of-doom/15417)

wh5929 replied:
> > 해당 코드가 다른 곳에서 재사용되는 경우가 없고 캡슐화/추상화하려는 대상의 수가 적어서,
> > 캡슐화/추상화하게 되면 결국 같은 기능을 하기 위한 코드 길이의 총합은 비슷할것 같습니다.
> > 코드자체는 한눈에 들어오기 때문에 가독성은 좋다고 생각하구요.
> 
> 캡슐화란 객체의 속성과 행위를 하나로 묶고, 실제 구현 내용 일부를 내부에 감추어 은닉하는 것을 말합니다.
> 
> 현재 이야기하고 있는 RunningGuideTabItem의 body에서는 ui로직을 모두 보여주고 있죠. 물론 연산자를 통해서 처리한것이기때문에 선언형 프로그래밍이 아니다라는건 아니지만 기본적으로 선언형 프로그래밍이 추구하는 건 로직의 절차를 숨기는 거라고 저는 생각하는데
> 
> 그런 측면에서 body에서 각각의 UI요소들을 구성하는 로직을 보여주는건 선언형이 추구하는 것과 다르게, 코드를 보는 입장에서 인지적으로 부하를 준다고 생각듭니다.
> 
> 지금은 코드를 직접 구현한지 얼마 안되셨으니 눈에 들어오시겠지만, 이걸 구현하지 않은 사람 또는 구현 후 일정시간이 지난뒤보면 body부분에서 ForEach 반복문을 통해 무언가를 반복적으로 만들어주고, 수평으로 image와 vstack을 배치하고, vstack에서 다시 타이틀과 서브타이틀을 구성하는 부분이라는걸 파악해야하죠.
> 
> 하지만 이걸 캡슐화를 통해서 표현하면 hstack에 썸네일과 타이틀스택만 구현하면 되죠. 물론 전체코드의 길이는 기존의 절차형방식이 더 짧습니다. 하지만 우리가 매번 RunningGuideTabItem을 수정할때 UI모든 로직을 볼필요가 없으니까요. body상에는 어떤 뷰들이 있고, 타이틀부분 또는 이미지 부분을 수정해야 한다면 해당 부분만 확인 후 수정할 수 있게되므로 캡슐화를 하는겁니다.
> 
> 다시 정리해서 말씀드리면 1.코드는 수정될 가능성이 크다. 2.코드를 다시 수정할때 코드를 보고 이해해야한다. 3.코드를 볼때 한번에 많은 정보가 들어오면 인지적 부하가 심하다. 4.이를 위해서 정보를 숨기고, 범주화(카테고리화)를 통해 이해를 돕고, 수정을 해당 파트만 진행할 수 있게 한다.
> 
> 추가적으로 말씀드리면 이러한 캡슐화, 추상화가 없다면 지금 저희가 사용하는 swiftui 프레임워크가 없고 매번 ui를 그릴때마다 지금 간단하게 사용하고 있는 연산자 또는 image text와 같은 객체를 반복적으로 계속 구현해줘야 됩니다.
> 
> 마지막으로 코드자체에 가독성이 좋다고 주관적으로 판단하는건 위험하다 생각합니다. 업무상 이메일, 문서를 작성할때 그 가독성을 판단하는건 작성자가 아니라 그걸 보는 사람이 하는거고, 실제로 가독성 좋은 코드를 작성하는 사람은 정말 극소수입니다. 저희가 현재 하고 있는 이 모든 과정자체가 가독성 좋은 코드를 짜기 위해 연습하는거고요. 저희가 단번에 그런 수준에 오를 수 있다고 생각하는건 정말 위험합니다.

자세한 설명에 감사드립니다. 가능한한 캡슐화, 추상화 사용하도록 하겠습니다.

DalioKim commented on GatheRunner/Presentaion/Running/RunningView.swift:
```diff
@@ -17,21 +17,21 @@ struct RunningView: View {
     var body: some View {
         VStack(alignment: .center) {
             Text("Header View")
```
[의견] HeaderView 구현이 완료되었는데 더미로 임시 처리한 부분을 기존의 구현완성된 부분과 연결하는게 좋을 것 같습니다.

wh5929 replied:
> [의견] HeaderView 구현이 완료되었는데 더미로 임시 처리한 부분을 기존의 구현완성된 부분과 연결하는게 좋을 것 같습니다.

네. 헤더뷰 포함했습니다~

DalioKim commented on GatheRunner/Presentaion/Running/RunningView.swift:
```diff
@@ -17,21 +17,21 @@ struct RunningView: View {
+            .frame(width: UIScreen.main.bounds.size.width, height: UIScreen.getHeightby(ratio: 0.0355), alignment: .leading)
             .frame(maxWidth: .infinity, maxHeight: 30, alignment: .leading)
 
             if justStartButtonIsSelected {
```
[의견] 조건에 따른 뷰를 보여주는 부분은 함수로 따로 캡슐화해주시는게 좋을 것 같습니다.

wh5929 replied:
> [의견] 조건에 따른 뷰를 보여주는 부분은 함수로 따로 캡슐화해주시는게 좋을 것 같습니다.

캡슐화했습니다~

DalioKim commented on GatheRunner/Presentaion/Running/RunningView.swift:
```diff
@@ -17,21 +17,21 @@ struct RunningView: View {
             }
-            .frame(width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height * 0.0355, alignment: .leading)
+            .frame(width: UIScreen.main.bounds.size.width, height: UIScreen.getHeightby(ratio: 0.0355), alignment: .leading)
             .frame(maxWidth: .infinity, maxHeight: 30, alignment: .leading)
```
[질문] 왜 frame 연산자를 2번 사용하신건가요?

wh5929 replied:
당시에는 이유가 있어서 썼는데 이유가 기억이 안나네요.
지금 테스트해보니 필요없는 코드로 확인되어 제거했습니다.

DalioKim commented on GatheRunner/Presentaion/Running/JustStartView.swift:
```diff
@@ -14,7 +14,7 @@ struct JustStartView: View {
     @StateObject private var manager = LocationManager()
     @Binding var selected: Int
     let timer = Timer.publish(every: 5.5, on: .main, in: .default).autoconnect()
```
[의견] timer 변수가 외부에서 접근할 일이 있을까요? 없다면 접근제한설정을 하는게 좋을 것 같습니다.

wh5929 replied:
그렇네요. private 사용하도록 수정할게요.

wh5929 replied:
> [의견] timer 변수가 외부에서 접근할 일이 있을까요? 없다면 접근제한설정을 하는게 좋을 것 같습니다.

그렇네요. private 사용하도록 수정할게요.

DalioKim commented:
[의견] RunningView의 .ignoresSafeArea(edges: .top) 연산자로 인해 상단의 카테고리버튼이 안됩니다.

DalioKim commented:
[질문] JustStartView mapView의 disabled 설정은 의도하신건가요?

DalioKim commented:
[의견] 네이밍 관련해서, 말씀드린 부분있지만 featuredTabView, JustStartView, RunGuideExperienceItemVM, FloatingView의 네이밍이 뭘 나타내는지 이해하기 어렵습니다.

DalioKim commented on GatheRunner/Presentaion/Running/JustStartView.swift:
```diff
@@ -36,25 +36,18 @@ struct JustStartView: View {
             .onTapGesture {
-                // timer의 connect 인스턴스 메서드는 upstream 내부에 정의되어 있으므로 upstream에 접근한다.
                 timer.upstream.connect().cancel()
             }
```
[질문] 러닝가이드의 아이템을 클릭하면 해당정보를 나타내는 상세화면으로 이동해야하는데 이 부분은 onTapGesture에서 어떤식으로 구현하실 예정이신가요?

wh5929 replied:
> [질문] 러닝가이드의 아이템을 클릭하면 해당정보를 나타내는 상세화면으로 이동해야하는데 이 부분은 onTapGesture에서 어떤식으로 구현하실 예정이신가요?

아직 어떻게 구현해야 할지 모르는 상태라, 구현할때 고민해봐야 할 것 같습니다.

wh5929 commented:
> [질문] JustStartView mapView의 disabled 설정은 의도하신건가요?

네. disable 해야 사용자가 맵을 컨트롤하는 문제를 막을 수 있습니다.

wh5929 commented:
> [의견] 네이밍 관련해서, 말씀드린 부분있지만 featuredTabView, JustStartView, RunGuideExperienceItemVM, FloatingView의 네이밍이 뭘 나타내는지 이해하기 어렵습니다.

직관적인 이름으로 바꿔볼게요.

DalioKim commented:
추상화와 관련된 글입니다.
https://dokhakdubini.tistory.com/395

DalioKim commented:
> > [질문] JustStartView mapView의 disabled 설정은 의도하신건가요?
> 
> 네. disable 해야 사용자가 맵을 컨트롤하는 문제를 막을 수 있습니다.

Ux적으로도 고려하신 부분 굉장히 좋네요

wh5929 commented:
> [의견] 네이밍 관련해서, 말씀드린 부분있지만 featuredTabView, JustStartView, RunGuideExperienceItemVM, FloatingView의 네이밍이 뭘 나타내는지 이해하기 어렵습니다.

featuredTabView와 FloatingView 와 같이 추상적인 네이밍은 수정하겠습니다.
JustStartView, RunGuideExperienceVM은 바꿀 필요없어 보입니다.
바로시작 뷰, 러닝가이드체험 뷰모델과 같이 콩글리쉬로 읽기 쉽다고 생각합니다.

wh5929 committed:
### FloatingView
1. FloatingView -> BottomButtonsView로 이름 변경
2. BottomButtonsView 내의 버튼들을 Extract하여 서브뷰로 구성
3. 리터럴값들을 상수로 선언

### JustStartView
1. 바인딩 변수명 변경 selected -> selectedRunGuideTabItem
2. featuredTabView 서브뷰 이름 변경 -> RunGuideTabView
3. RunGuideTabView 를 struct 를 이용해 외부에 선언
4. 리터럴값들을 상수로 선언
5. timer 상수를 private으로 접근제한자 설정

### RunGuideExperienceTabItem
1. 각 뷰들과 modifier들을 Extract -> 서브뷰로 만듬
2. 리터럴값들을 상수로 선언

### RunningView
1. if 문을 통해 뷰를 선택하는 부분을 캡슐화
2. 리터럴값들을 상수로 선언

### RunGuideExperienceItemVM
1. 이름 변경 : RunGuideExperienceItemVM -> RunGuideExperienceVM

DalioKim commented:
[의견]
PR이 너무 길어지고 리팩토링의 범위가 다양해서 우선 이슈에 등록해 남겨두고 머지하는게 좋을 것 같습니다.
RunGuideExperience이 없고, RunGuideExperienceVM과 파일명이 다른 부분만 확인 및 해결해주세요.

[이슈]

**네이밍**

1) 의견이 엇갈리고 있는데 JustStartView의 경우 직역하면 바로시작 화면 인데 UI상에 버튼라벨(바로시작버튼)로 사용될 이름 같습니다. 이보다는 그냥 시작화면 또는 러닝시작화면의 의미를 담은 네이밍이 어떨까요? 

2) RunGuideTabView의 selectedRunGuideTabItem, tabViewFrameWidthRatio와 같이 타입을 통해 유추할수 있는 의미(RunGuideTabView.selectedRunGuideTabItem와 같이 됩니다.)를 변수에서 중복적으로 사용하고 있는 경우가 있는데 이 경우도 selectedTabItem, WidthRatio와 같이 중복된 부분을 삭제해서 사용하는건 어떠신가요?

[참고링크](https://betterprogramming.pub/naming-conventions-in-swift-4b7ca5eed4d2)

**유형**

1) View 타입 분리:
제가 이해하기로는 swift UI에서 View는 하나의 타입인데, 러닝뷰의 각 요소들에 너무 많은 타입으로 분리되어있는 것 같습니다. 타입을 구분하는 기준에 대해서는 계속 공부하고 있긴한데, 제가 알고 있기로는 일반적으로 다른 역할을 담당하고 있고, 다른 이유로 코드를 변경해야 할 때 분리해야하는 걸로 알고 있습니다. RunningView의 BtnSelector, MenuBtn, BottomButtonsView 그리고 BottomButtonsView의 StartBtn, SideBtn 등 같은 뷰모델에 의존하고 있거나, 뷰모델이 없거나, 단순한 뷰로 구성되어있어 굳이 다른 타입의 View가 아닌 변수로 선언하는게 어떨까요?

    질문: RunningGuideView는 UI로직과 기능이 추가되는 뷰라고 하신거 맞죠?

[참고링크](https://swiftwithmajid.com/2019/10/30/view-composition-in-swiftui/)
[참고링크2](https://levelup.gitconnected.com/create-custom-view-components-in-swiftui-845b65e8ba94)


2) 네임스페이스:
사이즈와 관련된 상수들을 enum을 사용해서 정의하는건 어떨까요? 일반적으로 많이 사용하고 있는걸로 알고 있습니다.

[참고링크](https://www.jessesquires.com/blog/2015/07/19/swift-namespaced-constants/)

**가독성**

1) RunGuideTabView의 body 부분에 제스쳐와 관련된 UI로직이 있는데 이부분을 캡슐화하는 등 body 부분에서 보여질 로직들을 숨기는건 어떨까요? 제스쳐 캡슐화하는 부분은 제가 만들어서 올려보겠습니다.

wh5929 committed:
Merge branch 'develop' into feature/RV2022082801
